### PR TITLE
Add ES-module support and improve package building

### PR DESCRIPTION
**Pull request by szimeth-jozef**
I ran into an obstacle described in this issue #66, so I came up with a solution. I implemented bundling for multiple module types
- Added ES-modules
- Kept AMD modules (for backward compatibility reasons and I noticed that azure-devops-extension-api depends on this package/library)
- Bundled IIFE for browser use

by replacing the build script with rollup module bundler.

Other changes:
- Upgraded typescript version
- Added missing types
- Repalced vars with const/let to improve code consistency

### CODE REVIEW

**Comment by ChronosMasterOfAllTime** [comment]
It seems we have 3 PR's targeting the same issue:

#74 
#56 

I would vote for this PR since it generates support for multiple module types.

File: src/SDK.ts [hunk 1]
```diff
@@ -321,7 +321,7 @@ export async function getService<T>(contributionId: string): Promise<T> {
 * @param instance - Either: (1) an object instance, or (2) a function that takes optional context data and returns an object instance.
 */
 export function register<T = any>(instanceId: string, instance: T): void {
```
Comment (ChronosMasterOfAllTime) [inline src/SDK.ts:323]:
```suggestion
export function register<T extends Object>(instanceId: string, instance: T): void {

```

File: src/SDK.ts [hunk 2]
```diff
@@ -321,7 +321,7 @@ export async function getService<T>(contributionId: string): Promise<T> {
 * @param instance - Either: (1) an object instance, or (2) a function that takes optional context data and returns an object instance.
 */
 export function register<T = any>(instanceId: string, instance: T): void {
-    parentChannel.getObjectRegistry().register(instanceId, instance);
+    parentChannel.getObjectRegistry().register(instanceId, instance as Object);
```
Comment (ChronosMasterOfAllTime) [inline src/SDK.ts:324]:
```suggestion
    parentChannel.getObjectRegistry().register(instanceId, instance);

```
with the above suggestion this cast is unnecessary.

File: tsconfig.base.json [hunk 3]
```diff
@@ -1,20 +1,13 @@
...
+        "rootDir": "src/",
+        "experimentalDecorators": true,
+        "moduleResolution": "node",
+        "noImplicitAny": true,
+        "noImplicitThis": true,
+        "strict": true,
+        "lib": ["es5", "es6", "dom"]
+    },
+    "exclude": ["node_module/*"]
 }
```
Comment (ChronosMasterOfAllTime) [inline tsconfig.base.json:13]: new line EOF

File: tsconfig.esm.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+    "extends": "./tsconfig.base.json",
+    "compilerOptions": {
+        "target": "ES2022",
+        "module": "ESNext",
+        "declaration": true,
+        "declarationDir": "dist/esm"
+    }
+}
```
Comment (ChronosMasterOfAllTime) [inline tsconfig.esm.json:9]: new line EOF

File: .gitignore [hunk 5]
```diff
@@ -1,3 +1,3 @@
 npm-debug.log
 node_modules/
-bin/
\ No newline at end of file
+dist/
```
Comment (ChronosMasterOfAllTime) [inline .gitignore:3]: new line EOF

**Review by privatebay1 — APPROVED** [review]

File: tsconfig.base.json (hunk 3 above)
Comment (lohitakshgupta) [reply tsconfig.base.json:13]: Cherry-picking changes from here. Keeping `es2015.promise`

**Pull request closed without merging**